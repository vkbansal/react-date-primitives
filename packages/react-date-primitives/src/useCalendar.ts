import { useState, Dispatch, SetStateAction, useCallback } from 'react';

import {
  startOfMonth,
  getDaysOfMonth,
  Month,
  DayName,
  addMonths,
  setMonth,
  setYear
} from './utils';

export interface UseCalendarReturn extends Month {
  readonly month: Date;
  setMonth: Dispatch<SetStateAction<Date>>;
  setStartOfWeek: Dispatch<SetStateAction<DayName>>;
  nextMonth(): void;
  prevMonth(): void;
  setMonthOnly(month: number): void;
  setYearOnly(year: number): void;
}

export function useCalendar(month = new Date(), weekStartsOn?: DayName): UseCalendarReturn {
  const [currentMonth, setCurrentMonth] = useState(startOfMonth(month));
  const [startOfWeek, setStartOfWeek] = useState(weekStartsOn || DayName.SUNDAY);
  const { days, daysOfWeek } = getDaysOfMonth(currentMonth, startOfWeek);

  const nextMonth = useCallback(() => {
    setCurrentMonth((month) => startOfMonth(addMonths(month, 1)));
  }, []);

  const prevMonth = useCallback(() => {
    setCurrentMonth((month) => startOfMonth(addMonths(month, -1)));
  }, []);

  const setMonthOnly = useCallback((m: number) => {
    setCurrentMonth((month) => startOfMonth(setMonth(month, m)));
  }, []);

  const setYearOnly = useCallback((year: number) => {
    setCurrentMonth((month) => startOfMonth(setYear(month, year)));
  }, []);

  return {
    days,
    daysOfWeek,
    setMonth: setCurrentMonth,
    month: currentMonth,
    setStartOfWeek,
    nextMonth,
    prevMonth,
    setMonthOnly,
    setYearOnly
  };
}
