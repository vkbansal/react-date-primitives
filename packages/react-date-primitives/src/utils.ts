export const DayOfMonthSymbol = Symbol('DayOfMonth');
export const DayOfRangeMonthSymbol = Symbol('DayOfRangeMonth');

export enum DayName {
    MONDAY = 'MONDAY',
    TUESDAY = 'TUESDAY',
    WEDNESDAY = 'WEDNESDAY',
    THURSDAY = 'THURSDAY',
    FRIDAY = 'FRIDAY',
    SATURDAY = 'SATURDAY',
    SUNDAY = 'SUNDAY'
}

export const DaysOfTheWeek: DayName[] = [
    DayName.SUNDAY,
    DayName.MONDAY,
    DayName.TUESDAY,
    DayName.WEDNESDAY,
    DayName.THURSDAY,
    DayName.FRIDAY,
    DayName.SATURDAY
];

type Mutable<T> = {
    -readonly [P in keyof T]: T[P];
};

export interface DayOfMonth {
    readonly __type: typeof DayOfMonthSymbol;
    readonly dateObj: Date;
    readonly dayName: DayName;
    readonly inCurrentMonth: boolean;
    readonly ISODateString: string;
}

export interface DayOfRangeMonth extends Omit<DayOfMonth, '__type'> {
    readonly __type: typeof DayOfRangeMonthSymbol;
    readonly inRange: boolean;
    readonly isStart: boolean;
    readonly isEnd: boolean;
}

export interface Month {
    readonly days: DayOfMonth[];
    readonly daysOfWeek: DayName[];
    readonly month: Date;
}

export interface RangeMonth {
    readonly days: DayOfRangeMonth[];
    readonly month: Date;
}

export interface RangeMonths {
    readonly months: RangeMonth[];
    readonly daysOfWeek: DayName[];
}

const monthsCache = new Map<string, Month>();
const rangeMonthsCache = new Map<string, RangeMonths>();

export function addDays(date: Date, days: number): Date {
    return setDay(date, date.getDate() + days);
}

export function addMonths(date: Date, months: number): Date {
    return setMonth(date, date.getMonth() + months);
}

export function setDay(date: Date, day: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(day);

    return newDate;
}

export function setMonth(date: Date, month: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setMonth(month);

    return newDate;
}

export function setYear(date: Date, year: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(year);

    return newDate;
}

export function startOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(1);
    newDate.setHours(0, 0, 0, 0);

    return newDate;
}

export function endOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(date.getFullYear(), date.getMonth() + 1, 0);
    newDate.setHours(23, 59, 59, 999);

    return newDate;
}

export function isSameMonth(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth()
    );
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth() &&
        dateLeft.getDate() === dateRight.getDate()
    );
}

export function isDayAfter(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() > newDateToCompare.getTime();
}

export function isDayBefore(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() < newDateToCompare.getTime();
}

export function toISODateString(date: Date): string {
    const year = date.getFullYear();
    const month = (date.getMonth() + 1).toString().padStart(2, '0');
    const day = (date.getDate() + 0).toString().padStart(2, '0');

    return `${year}-${month}-${day}`;
}

export function callIfExists<T extends unknown[], U = unknown>(
    func: ((...args: T) => U) | undefined | null,
    ...args: T
): U | undefined {
    if (typeof func === 'function') {
        return func(...args);
    }
}

export function getDaysOfMonth(month: Date, weekStartsOn = DayName.SUNDAY): Month {
    const firstDate = startOfMonth(month);
    const dayOffset = DaysOfTheWeek.findIndex((d) => d === weekStartsOn);
    const key = `${toISODateString(firstDate)}-${weekStartsOn}`;

    if (dayOffset < 0) {
        throw new RangeError(`Invalid value for day provided: ${weekStartsOn}`);
    }

    const data = monthsCache.get(key);

    if (data) {
        return data;
    }

    const firstDayInCurrentMonth = firstDate.getDay();

    const days: Month['days'] = Array.from(
        { length: 6 * 7 },
        (_1, i): DayOfMonth => {
            const day = addDays(
                firstDate,
                i -
                    firstDayInCurrentMonth -
                    (dayOffset > firstDayInCurrentMonth ? 7 : 0) +
                    dayOffset
            );

            return {
                __type: DayOfMonthSymbol,
                dateObj: day,
                dayName: DaysOfTheWeek[(i + dayOffset) % 7],
                inCurrentMonth: isSameMonth(firstDate, day),
                ISODateString: toISODateString(day)
            };
        }
    );

    const daysOfWeek = DaysOfTheWeek.slice(dayOffset).concat(DaysOfTheWeek.slice(0, dayOffset));

    monthsCache.set(key, { days, daysOfWeek, month: firstDate });

    return { days, daysOfWeek, month: firstDate };
}

export function getDaysOfRangeMonth(
    rangeMonths: Date[],
    startDate?: Date | null,
    endDate?: Date | null,
    weekStartsOn = DayName.SUNDAY
): RangeMonths {
    let key = rangeMonths
        .map((month): string => {
            if (!(month instanceof Date)) {
                throw new TypeError('Given month is not an instance of Date');
            }
            return `${month.getFullYear()}-${month.getMonth()}`;
        })
        .join('|');

    key += `|${startDate ? toISODateString(startDate) : ''}`;
    key += `|${endDate ? toISODateString(endDate) : ''}`;
    key += `|${weekStartsOn}`;

    if (rangeMonthsCache.has(key)) {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return rangeMonthsCache.get(key)!;
    }

    const months = rangeMonths.reduce<Mutable<RangeMonths>>(
        (accumulator, currentMonth, i) => {
            const { days, daysOfWeek } = getDaysOfMonth(currentMonth, weekStartsOn);

            if (i === 0) {
                accumulator.daysOfWeek = daysOfWeek.slice(0);
            }

            const daysOfCurrentMonth = days.map(
                (day): DayOfRangeMonth => {
                    const isStart = Boolean(startDate && isSameDay(day.dateObj, startDate));
                    const isEnd = Boolean(endDate && isSameDay(day.dateObj, endDate));

                    return {
                        ...day,
                        __type: DayOfRangeMonthSymbol,
                        inRange:
                            isStart ||
                            isEnd ||
                            Boolean(
                                startDate &&
                                    isDayAfter(day.dateObj, startDate) &&
                                    endDate &&
                                    isDayBefore(day.dateObj, endDate)
                            ),
                        isStart,
                        isEnd
                    };
                }
            );

            accumulator.months.push({
                days: daysOfCurrentMonth,
                month: startOfMonth(currentMonth)
            });

            return accumulator;
        },
        { months: [], daysOfWeek: [] }
    );

    rangeMonthsCache.set(key, months);

    return months as RangeMonths;
}
