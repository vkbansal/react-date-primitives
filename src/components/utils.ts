import { DayOfMonthSubset } from './types';

const monthsCache: Map<string, DayOfMonthSubset[][]> = new Map();

export function setDay(date: Date, day: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(day);

    return newDate;
}

export function addDays(date: Date, days: number): Date {
    return setDay(date, date.getDate() + days);
}

export function setMonth(date: Date, month: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setMonth(month);

    return newDate;
}

export function addMonths(date: Date, months: number): Date {
    return setMonth(date, date.getMonth() + months);
}

export function setYear(date: Date, year: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(year);

    return newDate;
}

export function startOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(1);
    newDate.setHours(0, 0, 0, 0);

    return newDate;
}

export function endOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(date.getFullYear(), date.getMonth() + 1, 0);
    newDate.setHours(23, 59, 59, 999);

    return newDate;
}

export function isSameMonth(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth()
    );
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth() &&
        dateLeft.getDate() === dateRight.getDate()
    );
}

export function isDayAfter(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() > newDateToCompare.getTime();
}

export function isDayBefore(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() < newDateToCompare.getTime();
}

export function callIfExists(callback?: any, ...args: any[]) {
    if (typeof callback === 'function') {
        callback(...args);
    }
}

export function getDatesofMonth(month: Date): DayOfMonthSubset[][] {
    const firstDate = startOfMonth(month);
    const key = `${firstDate.getFullYear()}-${firstDate.getMonth()}`;

    const data = monthsCache.get(key);

    if (data) {
        return data;
    }

    const firstDayInCurrentMonth = firstDate.getDay();

    const newdata: DayOfMonthSubset[][] = Array.from(
        { length: 6 },
        (_1, i): DayOfMonthSubset[] => {
            const week = Array.from(
                { length: 7 },
                (_2, j): DayOfMonthSubset => {
                    const day = addDays(firstDate, i * 7 + j - firstDayInCurrentMonth);

                    return { date: day, inCurrentMonth: isSameMonth(firstDate, day) };
                }
            );

            return week;
        }
    );

    monthsCache.set(key, newdata);

    return newdata;
}
