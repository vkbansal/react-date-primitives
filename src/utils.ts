export interface DayOfMonth {
    date: Date;
    inCurrentMonth: boolean;
}

export type DaysOfMonth = DayOfMonth[][];

const monthsCache: Map<string, DaysOfMonth> = new Map();

export function setDay(date: Date, day: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(day);

    return newDate;
}

export function addDays(date: Date, days: number): Date {
    return setDay(date, date.getDate() + days);
}

export function setMonth(date: Date, month: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setMonth(month);

    return newDate;
}

export function addMonths(date: Date, months: number): Date {
    return setMonth(date, date.getMonth() + months);
}

export function setYear(date: Date, year: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(year);

    return newDate;
}

export function startOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(1);
    newDate.setHours(0, 0, 0, 0);

    return newDate;
}

export function endOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(date.getFullYear(), date.getMonth() + 1, 0);
    newDate.setHours(23, 59, 59, 999);

    return newDate;
}

export function isSameMonth(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth()
    );
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth() &&
        dateLeft.getDate() === dateRight.getDate()
    );
}

export function isDayAfter(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() > newDateToCompare.getTime();
}

export function isDayBefore(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() < newDateToCompare.getTime();
}

export function callIfExists(callback?: any, ...args: any[]) {
    if (typeof callback === 'function') {
        callback(...args);
    }
}

export function getDaysOfMonth(month: Date): DayOfMonth[][] {
    const firstDate = startOfMonth(month);
    const key = `${firstDate.getFullYear()}-${firstDate.getMonth()}`;

    const data = monthsCache.get(key);

    if (data) {
        return data;
    }

    const firstDayInCurrentMonth = firstDate.getDay();

    const newdata: DaysOfMonth = Array.from(
        { length: 6 },
        (_1, i): DayOfMonth[] => {
            const week = Array.from(
                { length: 7 },
                (_2, j): DayOfMonth => {
                    const day = addDays(firstDate, i * 7 + j - firstDayInCurrentMonth);

                    return { date: day, inCurrentMonth: isSameMonth(firstDate, day) };
                }
            );

            return week;
        }
    );

    monthsCache.set(key, newdata);

    return newdata;
}
