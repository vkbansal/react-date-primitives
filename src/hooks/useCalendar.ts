import { useState } from 'react';

import { startOfMonth, getDaysOfMonth, DayOfMonth } from '../utils';

export interface UseCalendar {
    days: DayOfMonth[][];
    setMonth(date: Date): void;
    month: Date;
}

export function useCalendar(month = new Date()): UseCalendar {
    const [currentMonth, setCurrentMonth] = useState(startOfMonth(month));
    const days = getDaysOfMonth(currentMonth);

    return {
        days,
        setMonth(date: Date) {
            setCurrentMonth(date);
        },
        month: currentMonth
    };
}
