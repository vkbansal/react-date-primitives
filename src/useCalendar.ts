import { useState } from 'react';

import { startOfMonth, getDaysOfMonth, DaysOfMonth, DayOfWeek } from './utils';

export interface Calendar extends DaysOfMonth {
    readonly month: Date;
    setMonth(date: Date): void;
}

export function useCalendar(
    /* istanbul ignore next */ month = new Date(),
    weekStartsOn?: DayOfWeek
): Calendar {
    const [currentMonth, setCurrentMonth] = useState(startOfMonth(month));
    const { days, daysOfWeek } = getDaysOfMonth(currentMonth, weekStartsOn);

    return {
        days,
        daysOfWeek,
        setMonth(date: Date): void {
            setCurrentMonth(date);
        },
        month: currentMonth
    };
}
