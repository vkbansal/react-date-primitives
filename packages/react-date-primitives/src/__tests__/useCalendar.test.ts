import { renderHook, act } from '@testing-library/react-hooks';

import { useCalendar } from '../useCalendar';
import { isSameMonth, DayName } from '../utils';

import MonthSerializer from './MonthSerializer';

expect.addSnapshotSerializer(MonthSerializer);

const YEAR = 2021;

describe('useCalendar hook tests', () => {
  test('gives correct days & month', () => {
    const date = new Date(YEAR, 0 /* Jan */, 1, 0, 0, 0, 0);
    const { result } = renderHook(() => useCalendar(date));

    expect(result.current).toMatchSnapshot();
    expect(isSameMonth(result.current.month, date)).toBe(true);
  });

  test('setMonth works', () => {
    const date1 = new Date(YEAR, 0 /* Jan */, 1, 0, 0, 0, 0);
    const date2 = new Date(YEAR, 3 /* Apr */, 1, 0, 0, 0, 0);
    const { result } = renderHook(() => useCalendar(date1));

    expect(isSameMonth(result.current.month, date1)).toBe(true);
    expect(result.current).toMatchSnapshot();

    act(() => {
      result.current.setMonth(date2);
    });
    expect(result.current).toMatchSnapshot();
    expect(isSameMonth(result.current.month, date2)).toBe(true);
  });

  test('setStartOfWeek', () => {
    const date = new Date(YEAR, 0 /* Jan */, 1, 0, 0, 0, 0);
    const { result } = renderHook(() => useCalendar(date));

    expect(result.current.daysOfWeek[0]).toBe(DayName.SUNDAY);
    expect(result.current).toMatchSnapshot();

    act(() => {
      result.current.setStartOfWeek(DayName.MONDAY);
    });

    expect(result.current.daysOfWeek[0]).toBe(DayName.MONDAY);
    expect(result.current).toMatchSnapshot();
  });
});
