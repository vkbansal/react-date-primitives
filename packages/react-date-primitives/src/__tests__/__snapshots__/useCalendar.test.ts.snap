// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCalendar hook tests gives correct days & month 1`] = `
|                Jan 2021                 |
|-----------------------------------------|
| SUN | MON | TUE | WED | THU | FRI | SAT |
|     |     |     |     |     |  01 |  02 |
|  03 |  04 |  05 |  06 |  07 |  08 |  09 |
|  10 |  11 |  12 |  13 |  14 |  15 |  16 |
|  17 |  18 |  19 |  20 |  21 |  22 |  23 |
|  24 |  25 |  26 |  27 |  28 |  29 |  30 |
|  31 |     |     |     |     |     |     |
`;

exports[`useCalendar hook tests setMonth works 1`] = `
|                Jan 2021                 |
|-----------------------------------------|
| SUN | MON | TUE | WED | THU | FRI | SAT |
|     |     |     |     |     |  01 |  02 |
|  03 |  04 |  05 |  06 |  07 |  08 |  09 |
|  10 |  11 |  12 |  13 |  14 |  15 |  16 |
|  17 |  18 |  19 |  20 |  21 |  22 |  23 |
|  24 |  25 |  26 |  27 |  28 |  29 |  30 |
|  31 |     |     |     |     |     |     |
`;

exports[`useCalendar hook tests setMonth works 2`] = `
|                Apr 2021                 |
|-----------------------------------------|
| SUN | MON | TUE | WED | THU | FRI | SAT |
|     |     |     |     |  01 |  02 |  03 |
|  04 |  05 |  06 |  07 |  08 |  09 |  10 |
|  11 |  12 |  13 |  14 |  15 |  16 |  17 |
|  18 |  19 |  20 |  21 |  22 |  23 |  24 |
|  25 |  26 |  27 |  28 |  29 |  30 |     |
|     |     |     |     |     |     |     |
`;

exports[`useCalendar hook tests setStartOfWeek 1`] = `
|                Jan 2021                 |
|-----------------------------------------|
| SUN | MON | TUE | WED | THU | FRI | SAT |
|     |     |     |     |     |  01 |  02 |
|  03 |  04 |  05 |  06 |  07 |  08 |  09 |
|  10 |  11 |  12 |  13 |  14 |  15 |  16 |
|  17 |  18 |  19 |  20 |  21 |  22 |  23 |
|  24 |  25 |  26 |  27 |  28 |  29 |  30 |
|  31 |     |     |     |     |     |     |
`;

exports[`useCalendar hook tests setStartOfWeek 2`] = `
|                Jan 2021                 |
|-----------------------------------------|
| MON | TUE | WED | THU | FRI | SAT | SUN |
|     |     |     |     |  01 |  02 |  03 |
|  04 |  05 |  06 |  07 |  08 |  09 |  10 |
|  11 |  12 |  13 |  14 |  15 |  16 |  17 |
|  18 |  19 |  20 |  21 |  22 |  23 |  24 |
|  25 |  26 |  27 |  28 |  29 |  30 |  31 |
|     |     |     |     |     |     |     |
`;
